README
NAME
    Task::Catmandu - Catmandu bundle

SYNOPSIS
      cpanm --interactive Task::Catmandu

      # Update cpanfile to most recent versions
      > perl -MTask::Catmandu -e cpanfile > cpanfile

AUTHOR
    Nicolas Steenlant <nicolas.steenlant@ugent.be>

CONTRIBUTORS
    Jakob Voss <jakob.voss@gbv.de>

LICENSE
    This library is free software; you can redistribute it and/or modify it
    under the same terms as Perl itself.

SEE ALSO
    <http://librecat.org/>

    The main code of this module is taken from Task::Plack.



Changes
Revision history for Perl extension Task::Catmandu

0.104  2014-10-20 11:12:28 CEST
        - migrate to Milla and update

0.103
        - bump versions

0.102
        - bump versions

0.101
        - Catmandu::FedoraCommons not default

0.1
        - Makefile.PL generator

0.0201
        - include Atom, XLS, BibTeX, Store::DBI

0.02
        - keep it simple

0.0102
        - include more packages

0.0101
        - update MANIFEST

0.01
        - initial release


Makefile.PL
use inc::Module::Install;

name 'Task-Catmandu';
all_from 'lib/Task/Catmandu.pm';
readme_from 'lib/Task/Catmandu.pm';
license 'perl';

feature 'repl (interactive environment)', -default => 1,
'Catmandu::Cmd::repl', '0.01',
;
feature 'MongoDB store', -default => 1,
'Catmandu::Store::MongoDB', '0.0302',
;
feature 'CouchDB store', -default => 1,
'Catmandu::Store::CouchDB', '0.01',
;
feature 'ElasticSearch store', -default => 1,
'Catmandu::Store::ElasticSearch', '0.0205',
;
feature 'Solr store', -default => 1,
'Catmandu::Store::Solr', '0.0204',
;
feature 'DBI store', -default => 1,
'Catmandu::Store::DBI', '0.04',
;
feature 'FedoraCommons store', -default => 0,
'Catmandu::FedoraCommons', '0.271',
;
feature 'MARC importer, exporter and fixes', -default => 1,
'Catmandu::MARC', '0.113',
;
feature 'bibliographic importers', -default => 1,
'Catmandu::PubMed', '0.01',
'Catmandu::ArXiv', '0.01',
'Catmandu::PLoS', '0.01',
'Catmandu::Inspire', '0.261',
;
feature 'Atom importer and exporter', -default => 1,
'Catmandu::Atom', '0.02',
;
feature 'OAI importer', -default => 1,
'Catmandu::OAI', '0.02',
;
feature 'SRU importer', -default => 1,
'Catmandu::SRU', '0.033',
;
feature 'Z39.50 importer', -default => 0,
'Catmandu::Z3950', '0.04',
;
feature 'Twitter importer', -default => 1,
'Catmandu::Importer::Twitter', '0.03',
;
feature 'LDAP importer', -default => 1,
'Catmandu::LDAP', '0.0103',
;
feature 'BibTeX exporter', -default => 1,
'Catmandu::BibTeX', '0.04',
;
feature 'XLS exporter', -default => 1,
'Catmandu::XLS', '0.03',
;
feature 'Dancer integration', -default => 0,
'Dancer::Session::Catmandu', '0.02',
'Dancer::Plugin::Catmandu::OAI', '0.0304',
'Dancer::Plugin::Catmandu::SRU', '0.02',
;

auto_install;
auto_set_repository;
WriteAll;



lib/Task/Catmandu.pm
package Task::Catmandu;

use strict;
use warnings;

our $VERSION = '0.104';

use ExtUtils::MakeMaker;

use base qw(Exporter);
our @EXPORT = qw(cpanfile);

sub dependencies {
    return (
        'Catmandu core', 1, [
            'Catmandu',
        ],
        'Useful additions', 1, [
            'Catmandu::Fix::cmd',
            'Catmandu::Fix::Date',
            'Catmandu::Cmd::repl',
            'Catmandu::Store::CHI',
            'Catmandu::Fix::get_json',
            'Catmandu::Exporter::Table',
            'Catmandu::Exporter::Template',
            'Catmandu::Serializer::storable', # ?
            'Catmandu::Validator::JSONSchema',
        ],
        'Data structuring languages', 1, [
            'Catmandu::XML',
            'Catmandu::RDF',
        ],
        'Databases', 0, [
            'Catmandu::Store::DBI',
            'Catmandu::Store::MongoDB',
            'Catmandu::Store::CouchDB',
        ],
        'Search servers', 0, [
            'Catmandu::Store::ElasticSearch',
            'Catmandu::Store::Solr',
            'Catmandu::Store::Lucy',
        ],
        'Bibliographic formats', 1, [
            'Catmandu::MARC',
            'Catmandu::PICA',
            'Catmandu::MAB2',
            'Catmandu::BibTeX',
            'Catmandu::MODS',
        ],
        'Bibliographic importers', 'bibliographic_databases', [
            'Catmandu::CrossRef',
            'Catmandu::PubMed',
            'Catmandu::ArXiv',
            'Catmandu::PLoS',
            'Catmandu::Inspire',
            'Catmandu::ArXiv',
            'Catmandu::OCLC',
        ],
        'FedoraCommons store', 'fedora', [
            'Catmandu::FedoraCommons',
        ],
        'Atom importer and exporter', 1, [
            'Catmandu::Atom',
        ],
        'OAI importer', 1, [
            'Catmandu::OAI',
        ],
        'SRU importer', 1, [
            'Catmandu::SRU',
        ],
        'Z39.50 importer', 'z3950', [
            'Catmandu::Z3950',
        ],
        'Twitter importer', 1, [
            'Catmandu::Importer::Twitter',
        ],
        'Wikidata importer', 1, [
            'Catmandu::Wikidata'
        ],
        'EuroPMC', 0, [
            'Catmandu::EuropePMC'
        ],
        'LDAP importer', 1, [
            'Catmandu::LDAP',
        ],
        'Excel .xls and .xlsx files', 1, [
            'Catmandu::XLS',
        ],
        'Aleph-X services', 'alephx', [
            'Catmandu::AlephX',
        ],
        'MediaMosa', 0, [
            'Catmandu::MediaMosa',
        ],
        'Activiti', 0, [
            'Catmandu::Activiti',
        ],
        'Perl recent releases', 'cpan', [
            'Catmandu::Importer::CPAN',
        ],
        'PSGI integration', 'psgi', [
            'Plack::Session::Store::Catmandu',
        ],
        'Dancer integration', 'dancer', [
            'Dancer::Session::Catmandu',
            'Dancer::Plugin::Catmandu::OAI',
            'Dancer::Plugin::Catmandu::SRU',
        ],
    );
}

sub iter_deps {
    my $cb = shift;
    my @deps = dependencies();
    while (my($name, $feature, $deps) = splice @deps, 0, 3) {
        $cb->($name, $feature, $deps);
    }
}

sub cpanfile {
    my $fh = shift || *STDOUT;

    iter_deps(sub {
        my($name, $feature, $modules) = @_;
        if ($feature eq 1) {
            $fh->print("# $name\n");
            for my $module (@$modules) {
                $fh->print("requires '$module', '", version_for($module), "',\n");
            }
        } else {
            (my $ident = $name) =~ s/[^A-Za-z_]+/_/g;
            $feature ||= lc($ident);
            $fh->print("feature '", $feature, "', '$name' => sub {\n");
            for my $module (@$modules) {
                $fh->print("  requires '$module', '", version_for($module), "',\n");
            }
            $fh->print("};\n");
        }
        $fh->print("\n");
    });
}

sub version_for {
    my $dist = shift;
    (my $module = $dist) =~ s/-/::/g;
    my $info = `cpanm --info $module ` or return;
    return ($info =~ /-([\d\.]+)\.tar\.gz/)[0];
}

1;

__END__

=encoding utf-8

=for stopwords

=head1 NAME

Task::Catmandu - Catmandu bundle

=head1 SYNOPSIS

  cpanm --interactive Task::Catmandu

  # Update cpanfile to most recent versions
  > perl -MTask::Catmandu -e cpanfile > cpanfile

=head1 AUTHOR

Nicolas Steenlant E<lt>nicolas.steenlant@ugent.beE<gt>

=head1 CONTRIBUTORS

Jakob Voss E<lt>jakob.voss@gbv.deE<gt>

=head1 LICENSE

This library is free software; you can redistribute it and/or modify
it under the same terms as Perl itself.

=head1 SEE ALSO

L<http://librecat.org/>

The main code of this module is taken from L<Task::Plack>.

=cut


generate_Makefile_PL.pl
#!/usr/bin/perl

# taken from Task::Plack
sub version_for {
    my $dist = shift;

    (my $module = $dist) =~ s/-/::/g;
    my $info = `cpanm --info $module ` or return;
    return ($info =~ /-([\d\.]+)\.tar\.gz/)[0];
}

my @features = (
    [ 'repl (interactive environment)', 1,
      ['Catmandu::Cmd::repl'],
    ],
    # stores
    [ 'MongoDB store', 1,
      ['Catmandu::Store::MongoDB'],
    ],
    [ 'CouchDB store', 1,
      ['Catmandu::Store::CouchDB'],
    ],
    [ 'ElasticSearch store', 1,
      ['Catmandu::Store::ElasticSearch'],
    ],
    [ 'Solr store', 1,
      ['Catmandu::Store::Solr'],
    ],
    [ 'DBI store', 1,
      ['Catmandu::Store::DBI'],
    ],
    [ 'FedoraCommons store', 0,
      ['Catmandu::FedoraCommons'],
    ],
    # importers
    [ 'MARC importer, exporter and fixes', 1,
      ['Catmandu::MARC'],
    ],
    [ 'bibliographic importers', 1,
      ['Catmandu::PubMed',
       'Catmandu::ArXiv',
       'Catmandu::PLoS',
       'Catmandu::Inspire'],
    ],
    [ 'Atom importer and exporter', 1,
      ['Catmandu::Atom'],
    ],
    [ 'OAI importer', 1,
      ['Catmandu::OAI'],
    ],
    [ 'SRU importer', 1,
      ['Catmandu::SRU'],
    ],
    [ 'Z39.50 importer', 0,
      ['Catmandu::Z3950'],
    ],
    [ 'Twitter importer', 1,
      ['Catmandu::Importer::Twitter'],
    ],
    [ 'LDAP importer', 1,
      ['Catmandu::LDAP'],
    ],
    # exporters
    [ 'BibTeX exporter', 1,
      ['Catmandu::BibTeX'],
    ],
    [ 'XLS exporter', 1,
      ['Catmandu::XLS'],
    ],
    # dancer integration
    [ 'Dancer integration', 0,
      ['Dancer::Session::Catmandu',
       'Dancer::Plugin::Catmandu::OAI',
       'Dancer::Plugin::Catmandu::SRU'],
    ],

);

print <<PERL;
use inc::Module::Install;

name 'Task-Catmandu';
all_from 'lib/Task/Catmandu.pm';
readme_from 'lib/Task/Catmandu.pm';
license 'perl';

PERL

for my $feature (@features) {
    my ($name, $default, $pkgs) = @$feature;
    print "feature '$name', -default => $default,\n";
    for my $pkg (@$pkgs) {
        my $version = version_for($pkg);
        print "'$pkg', '$version',\n";
    }
    print ";\n";
}

print <<PERL;

auto_install;
auto_set_repository;
WriteAll;

PERL



versions.pl
feature 'repl (interactive environment)', -default => 1,
  'Catmandu::Cmd::repl', '0.01',
;
feature 'MongoDB store', -default => 1,
  'Catmandu::Store::MongoDB', '0.0201',
;
feature 'ElasticSearch store', -default => 1,
  'Catmandu::Store::ElasticSearch', '0.0201',
;
feature 'Solr store', -default => 1,
  'Catmandu::Store::Solr', '0.02',
;
feature 'FedoraCommons store', -default => 1,
  'Catmandu::FedoraCommons', '0.21',
;
feature 'MARC importer and MARC fixes', -default => 1,
  'Catmandu::MARC', '0.03',
;
feature 'bibliographic importers', -default => 1,
  'Catmandu::PubMed', '0.01',
  'Catmandu::ArXiv', '0.01',
  'Catmandu::PLoS', '0.01',
;
feature 'OAI importer', -default => 1,
  'Catmandu::OAI', '0.01',
;
feature 'SRU importer', -default => 1,
  'Catmandu::SRU', '0.0101',
;
feature 'Z39.50 importer', -default => 0,
  'Catmandu::Z3950', '0.03',
;
feature 'Twitter importer', -default => 0,
  'Catmandu::Importer::Twitter', '0.02',
;
feature 'LDAP importer', -default => 0,
  'Catmandu::LDAP', '0.0102',
;
feature 'Dancer integration', -default => 0,
  'Dancer::Session::Catmandu', '0.02',
  'Dancer::Plugin::Catmandu::OAI', '0.0301',
  'Dancer::Plugin::Catmandu::SRU', '0.02',
;
